cmd/object-api-utils.go

```diff
@@ -597,3 +598,39 @@ func (g *GetObjectReader) Read(p []byte) (n int, err error) {
 	}
 	return
 }
+
+// PutObjectReader is a type that wraps sio.EncryptReader and
+// underlying hash.Reader in a struct
+type PutObjectReader struct {
+	DataReader *hash.Reader // actual data stream
+	OrigReader *hash.Reader // original data stream
+}
+
+// Size returns the absolute number of bytes the Reader
+// will return during reading. It returns -1 for unlimited
+// data.
+func (p *PutObjectReader) Size() int64 {
+	return p.DataReader.Size()
+}
+
+// NewPutObjectReader returns a new PutObjectReader and holds
+// reference to underlying data stream from client and the encrypted
+// data reader
+func NewPutObjectReader(r *hash.Reader) *PutObjectReader {
+	return &PutObjectReader{OrigReader: r, DataReader: r}
+}
```

Comment:
To really make the purpose of this new struct `PutObjectReader` clear, all instantiation of the type should happen via constructor functions like above. This particular function looks like it applies only for un-encrypted put data?

Reply:
It is initialized to same reader for both encrypted and plain text case. For encryption, the object encryption key is set on the OrigReader so that encrypted MD5Sum of client content can be fetched. For encryption, DataReader is changed to use the one returned by sio.EncryptReader